File: solutions/lua/meetup/1/meetup.lua
return function(dt)
	local order = { first = 1, second = 2, third = 3, fourth = 4 }
	local days_set = {}
	local days = {}
	for d = 1, 31 do
		local tm = os.time({ year = dt.year, month = dt.month, day = d })
		local leap_year = ((dt.year % 400) == 0) or (((dt.year % 4) == 0) and not ((dt.year % 100) == 0))
		local isvalid = not (dt.month == 2 and not leap_year and d > 28)
		if os.date("%A", tm) == dt.day and isvalid then
			days[#days + 1] = d
			days_set[d] = true
		end
	end
	if dt.week == "last" then
		return days[#days]
	elseif dt.week == "teenth" then
		for i = 13, 19 do
			if days_set[i] then return i end
		end
	end
	return days[order[dt.week]]
end

File: solutions/clojure/meetup/1/src/meetup.clj
(ns meetup
  (:import [java.time YearMonth LocalDate]))

(def ^:private weekdays
  (zipmap [:monday :tuesday :wednesday :thursday
           :friday :saturday :sunday]
          (range 1 8)))

(defn meetup [month year weekday week]
  (let [start (if (= week :last)
                (- (.lengthOfMonth (YearMonth/of year month)) 6)
                ({:first 1 :second 8 :third 15
                  :fourth 22 :teenth 13} week))
        offset (mod (- (weekdays weekday)
                       (->> (LocalDate/of year month start)
                            .getDayOfWeek
                            .getValue))
                    7)]
    [year month (+ start offset)]))
